### PR TITLE
Add set_native_scripts() to TransactionBuilder

### PR DESCRIPTION
**Pull request by siegfried**
This PR is to fix issue #400. `NativeScripts` is not only need by minting but also spending on Multi-Sig scripts.

### CODE REVIEW

**Comment by siegfried** [comment]
Also, I don't think you can fake a `NativeScripts` with the same size by mere `ScriptHash` [here](https://github.com/Emurgo/cardano-serialization-lib/blob/c0a8f9d232bfe97c845d94a00d10419ae422f67c/rust/src/tx_builder.rs#L99). However, when spending UTxOs from a `NativeScript`, the `NativeScript` is required to be included in the transaction, right? With `set_native_scripts()` added, the size should be able to be calculated [here](https://github.com/Emurgo/cardano-serialization-lib/blob/c0a8f9d232bfe97c845d94a00d10419ae422f67c/rust/src/tx_builder.rs#L119). We could just validate the presence of the `NativeScript` in `add_script_input()`.

**Comment by siegfried** [comment]
In case there might be an update which enables Cardano nodes caching scripts. I suggest to delete [this part](https://github.com/siegfried/cardano-serialization-lib/blob/44469b30e40d52587767cfbe6c0ffd5eaa095628/rust/src/tx_builder.rs#L96) and push the validation to the node.

**Comment by siegfried** [comment]
How about providing a way to let the user decide how many [VKeys](https://github.com/Emurgo/cardano-serialization-lib/blob/c0a8f9d232bfe97c845d94a00d10419ae422f67c/rust/src/tx_builder.rs#L82) should be used? You might be able to count the keys for native scripts but what about the Plutus scripts? Also for native scripts like `Any` and `KofN` the keys included in the transactions can be reduced to the minimum in order to reduce the fee and network loads.

File: rust/src/tx_builder.rs [hunk 1]
```diff
@@ -801,14 +806,14 @@ impl TransactionBuilder {
 
     /// Returns a copy of the current mint witness scripts in the builder
     pub fn get_mint_scripts(&self) -> Option<NativeScripts> {
```
Comment (siegfried) [inline rust/src/tx_builder.rs:808]: I'm not intended to break the existing interface but this could be `get_native_scripts` as well.

**Comment by siegfried** [comment]
FYI. We've manage to calculate the correct fee by this PR. We've done tests locally.

File: rust/src/tx_builder.rs [hunk 2]
```diff
@@ -93,14 +93,6 @@ fn fake_full_tx(tx_builder: &TransactionBuilder, body: TransactionBody) -> Resul
             Some(result)
         },
     };
-    let script_keys: Option<NativeScripts> = match tx_builder.input_types.scripts.len() {
-        0 => None,
-        _x => {
-            // TODO: figure out how to populate fake witnesses for these
```
Comment (siegfried) [inline rust/src/tx_builder.rs:99]: You can't.

**Review by vsubhuman — COMMENTED** [review]
/check

**Comment by vsubhuman** [comment]
Thank you for the suggested change, @siegfried ! But this PR will be closed in favour of the new PR: https://github.com/Emurgo/cardano-serialization-lib/pull/436 , which is supposed to add native scripts handling in a consistent way along with plutus scripts handling

**Pull request closed without merging**